#!/usr/bin/env nextflow

include { INITIAL_ORTHOGROUPS } from './subworkflows/initial_orthogroups'
//include { WF_CLUSTERING } from './subworkflows/wf_clustering'

workflow {
    // Initialize variables for messages
    def orthofinder_msg      = "ğŸ” OrthoFinder: SKIP âŒ"
    def broccoli_msg         = "ğŸ¥¦ Broccoli:    SKIP âŒ"
    def search_msg           = "ğŸ”¨ HMMSEARCH:   SKIP âŒ"
    def cluster_dmnd_mcl_msg = "ğŸ’ DIAMOND:     SKIP âŒ"
    def cluster_mmseqs_msg   = "ğŸš€ MMseqs2:     SKIP âŒ"
    def vs_msg               = "âš ï¸  Clustering ALL orthogroups  âš ï¸"

    // Check if at least one of OrthoFinder or Broccoli is set to run
    if (!params.run.orthofinder && !params.run.broccoli) {
        log.error """
        ============================================================
        Error: Neither OrthoFinder nor Broccoli is set to run!
        
        You must set at least one of the following to true:
        - params.run.orthofinder
        - params.run.broccoli
        
        Exiting the pipeline.
        ============================================================
        """
        exit 1
    }

    // Update messages based on parameters
    if (params.run.orthofinder) orthofinder_msg     = "ğŸ” OrthoFinder: RUN âœ…"
    if (params.run.broccoli)    broccoli_msg        = "ğŸ¥¦ Broccoli:    RUN âœ…"
    if (params.run.search) {
        vs_msg = "âš ï¸  Clustering ONLY HMMSEARCH results  âš ï¸"
        search_msg           = "ğŸ”¨ HMMSEARCH:  RUN âœ…"
    }
    if (params.run.cluster_dmnd_mcl) cluster_dmnd_mcl_msg = "ğŸ’ DIAMOND:    RUN âœ…"
    if (params.run.cluster_mmseqs)   cluster_mmseqs_msg   = "ğŸš€ MMseqs2:    RUN âœ…"

    // Print workflow messages
    log.info """
    Pipeline workflow that will be executed:
    ---------------------------
    ğŸ¥‡ Initial orthology:
        ${broccoli_msg}
        ${orthofinder_msg}

    ğŸ‘ª Gene family search of orthologs (SEARCH):
        ${search_msg}

    ğŸ‡ Clustering:
        ${vs_msg}
        ${cluster_dmnd_mcl_msg}
        ${cluster_mmseqs_msg}
    """

    // Run the orthology workflow
    INITIAL_ORTHOGROUPS(
        params.fasta_dir,
        params.orthofinder.prior_run,
        params.orthofinder.min_sequences,
        params.broccoli.args,
        params.broccoli.min_sequences,
        params.search.gene_family_info,
        params.search.gene_family_name,
        params.search.hmm_dir,
        params.outdir,
        params.run.orthofinder,
        params.run.broccoli,
        params.run.search
    )

    // Run the clustering workflow
    //WF_CLUSTERING(
    //    WF_ORTHOLOGY.out.search_results,
    //    params.cluster.dmnd.args,
    //    params.cluster.mcl.args,
    //   params.cluster.mcl.inflation,
    //    params.run.cluster_dmnd_mcl,
    //    params.run.cluster_mmseqs,
    //    params.run.search ? "searches" : "all"
    //)
}